src/minimetrics/core.py

```diff
@@ -0,0 +1,424 @@
+        self.sum += value
+
+    def serialize_value(self) -> Any:
+        # For now, we compress gauges with the last value.
+        return self.last
+
+
+class DistributionMetric(Metric[float]):
+    __slots__ = ("value",)
+
+    def __init__(self) -> None:
+        self.value: List[float] = []
+
+    @property
+    def current_complexity(self) -> int:
+        return len(self.value)
+
+    def add(self, value: float) -> None:
+        self.value.append(float(value))
+
+    def serialize_value(self) -> Any:
+        return self.value
+
+
+class SetMetric(Metric[Union[str, int]]):
+    __slots__ = ("value",)
+
+    def __init__(self) -> None:
+        self.value: Set[Union[str, int]] = set()
+
+    @property
+    def current_complexity(self) -> int:
+        return len(self.value)
+
+    def add(self, value: Union[str, int]) -> None:
+        self.value.add(value)
+
+    def serialize_value(self) -> Any:
+        def _hash(x: Any) -> int:
+            if isinstance(x, str):
+                return zlib.crc32(x.encode("utf-8")) & 0xFFFFFFFF
+            return int(x)
+
+        return [_hash(x) for x in self.value]
+
+
+METRIC_TYPES: Dict[str, Callable[[], Metric[Any]]] = {
+    "c": CounterMetric,
+    "g": GaugeMetric,
+    "d": DistributionMetric,
+    "s": SetMetric,
+}
+
+
+class Aggregator:
+    ROLLUP_IN_SECONDS = 10.0
+    MAX_COMPLEXITY = 100000
+
+    def __init__(self) -> None:
+        self.buckets: Dict[ComposedKey, Metric[Any]] = {}
+        self._bucket_complexity: int = 0
+        self._lock: Lock = Lock()
+        self._running: bool = True
+        self._flush_event: Event = Event()
+        self._force_flush: bool = False
+        # Thread handling the flushing loop.
+        self._flusher: Optional[Thread] = Thread(target=self._flush_loop)
+        self._flusher.daemon = True
+        self._flusher.start()
+
+    def _flush_loop(self) -> None:
+        # We check without locking these variables, such racy check can lead to problems if we are not careful. The most
+        # important invariant of the system that needs to be maintained is that if running and force_flush are false,
+        # the number of buckets is equal to 0.
+        while self._running or self._force_flush:
+            self._flush()
+            self._flush_event.wait(2.0)
+
+    def _flush(self):
+        with self._lock:
+            cutoff = time.time() - self.ROLLUP_IN_SECONDS
+            complexity_to_remove = 0
+            buckets = self.buckets
+            force_flush = self._force_flush
+            flushed_buckets = set()
+            extracted_metrics = []
+
+            for bucket_key, metric in buckets.items():
+                ts, ty, name, unit, tags = bucket_key
+                if not force_flush and ts > cutoff:
+                    continue
+
+                extracted_metric: ExtractedMetric = {
+                    "type": ty,
+                    "name": name,
+                    "value": metric.serialize_value(),
```

Comment:
There's more work than necessary performed inside the lock. Consider to just collect the buckets here, and then run `serialize_value`, ideally on a generator, directly in `_emit()`